### PR TITLE
add ability to check health in parent from the module

### PR DESCRIPTION
**Pull request by thekemkid**
Tests are known to be failing atm, I needed to park this for a bit before finishing it off.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -20,6 +25,11 @@ function underPressure (fastify, opts, next) {
   var lastCheck = now()
   const timer = setInterval(updateMemoryUsage, sampleInterval)
 
+  var externalsHealthy = false
+  const doCheck = () => healthCheck().then(externalHealth => { externalsHealthy = externalHealth })
+  doCheck()
```
Comment (mcollina) [inline index.js]: what if `healthCheck` throws?

File: index.js [hunk 2]
```diff
@@ -20,6 +25,11 @@ function underPressure (fastify, opts, next) {
...
+  const externalHealthCheckTimer = setInterval(doCheck, healthCheckInterval)
```
Comment (mcollina) [inline index.js]:
```suggestion
  const externalHealthCheckTimer = setInterval(doCheck, healthCheckInterval)
  externalHealthCheckTimer.unref()
```

File: index.js [hunk 3]
```diff
@@ -20,6 +25,11 @@ function underPressure (fastify, opts, next) {
   var lastCheck = now()
   const timer = setInterval(updateMemoryUsage, sampleInterval)
```
Comment (mcollina) [inline index.js]: I would `unref()` also this one.

File: test.js [hunk 4]
```diff
@@ -262,6 +262,74 @@ test('Expose custom status route', t => {
...
+        t.error(err)
+        t.strictEqual(response.statusCode, 200)
+        t.deepEqual(JSON.parse(body), {
+          hello: 'world'
+        })
+        fastify.close()
+      })
+    })
+  })
+})
```
Comment (mcollina) [inline test.js]:
```suggestion
})

```

File: index.js [hunk 5]
```diff
@@ -27,15 +25,23 @@ function underPressure (fastify, opts, next) {
...
+  if (healthCheck) {
+    assert(typeof healthCheck === 'function', 'opts.healthCheck should be a function that returns a promise that resolves to true or false')
+
+    const doCheck = () => healthCheck()
+      .then(externalHealth => { externalsHealthy = externalHealth })
+      .catch((error) => {
+        externalsHealthy = false
+        fastify.log.error('external healthCheck function suupplied to `under-pressure` threw an error. setting the service status to unhealthy.', { error })
+      })
+    doCheck()
```
Comment (mcollina) [inline index.js]: we need to wait the first run of this before calling `next`.

File: test.js [hunk 6]
```diff
@@ -262,6 +262,79 @@ test('Expose custom status route', t => {
...
+    fastify.get('/', (req, reply) => {
+      reply.send({ hello: 'world' })
+    })
+
+    fastify.listen(0, (err, address) => {
+      t.error(err)
+      fastify.server.unref()
+
+      // Increased to prevent Travis to fail
+      process.nextTick(() => sleep(1000))
```
Comment (mcollina) [inline test.js]: this is a code smell, why is this needed?

Comment (thekemkid) [reply test.js]: copy pasta from  previous tests. will remove.

File: index.js (hunk 5 above)
Comment (thekemkid) [reply index.js]: I've added a fix for this, but dislike the code I created because of it. The if  guard around the next() call at the end of the function seems out of place. I've added a comment there but other thoughts/feedback needed.

File: test.js [hunk 7]
```diff
@@ -329,6 +329,25 @@ test('Custom health check', t => {
...
+        called = true
+      }
+    })
+
+    fastify.listen(0, (err) => {
+      t.error(err)
+      t.true(called)
+      fastify.close()
+    })
+  })
```
Comment (mcollina) [inline test.js]: would this test fail without the fix in the code? You might have to add a `sleep` in the healthCheck.

Comment (thekemkid) [reply test.js]: I added a promisified wait so the sleep doesn't block the eventloop to allow other async components to initialise to ensure that it is just that check blocking starting

File: index.js [hunk 8]
```diff
@@ -1,6 +1,7 @@
 'use strict'
 
 const fp = require('fastify-plugin')
+const assert = require('assert')
 
 function underPressure (fastify, opts, next) {
```
Comment (mcollina) [inline index.js]: it's better if you rewrite this using an async function and remove next.. using an `await` for the health check, so you can avoid that bad comment in the end.

File: index.js [hunk 9]
```diff
@@ -19,6 +22,29 @@ function underPressure (fastify, opts, next) {
...
+  var externalHealthCheckTimer
+  if (healthCheck) {
+    assert(typeof healthCheck === 'function', 'opts.healthCheck should be a function that returns a promise that resolves to true or false')
+
+    const doCheck = async () => {
+      try {
+        externalsHealthy = await healthCheck()
+      } catch (error) {
+        externalsHealthy = false
+        fastify.log.error('external healthCheck function supplied to `under-pressure` threw an error. setting the service status to unhealthy.', { error })
```
Comment (mcollina) [inline index.js]: You should put { err } as the beginning of the call to .error to use the error serializer

**Comment by thekemkid** [comment]
@mcollina Can you re-review?

@delvedor do you have any comments on this as well?

I'd like to get this merged soon if possible :)

**Comment by mcollina** [comment]
I would like the healthcheck to be called every time the status route is called. Could you add that?

File: index.js [hunk 10]
```diff
@@ -1,14 +1,17 @@
...
-function underPressure (fastify, opts, next) {
+async function underPressure (fastify, opts) {
   opts = opts || {}
 
   const sampleInterval = opts.sampleInterval || 5
   const maxEventLoopDelay = opts.maxEventLoopDelay || 0
   const maxHeapUsedBytes = opts.maxHeapUsedBytes || 0
   const maxRssBytes = opts.maxRssBytes || 0
+  const healthCheck = opts.healthCheck || false
+  const healthCheckInterval = opts.healthCheckInterval || 1000
```
Comment (mcollina) [inline index.js]: Can you make this 0/no check as the default?

**Comment by thekemkid** [comment]
lgtm
